### ISSUE TITLE
Deadlock in EntityFrameworkQueryableExtensions.ToListAsync if used with synchronization context

### ISSUE THREAD

**Issue by Belorus** [dotnet/efcore#28159]
```
 public async Task<ActionResult> Index()
        {
            var ctx = GetDbContext();
            var result = ctx.Set<GetServicePlanConfigurationByCurrencyResult>().FromSqlRaw("select 1").ToListAsync().ConfigureAwait(false).GetAwaiter().GetResult();
            
            return View();
        }
```

This code deadlocks inside ToListAsync()

**Comment by ErikEJ** [comment]
Why not just use:

    var result = await ctx.Set<GetServicePlanConfigurationByCurrencyResult>().FromSqlRaw("select 1").ToListAsync();

**Comment by Belorus** [comment]
> Why not just use:
> 
> ```cs
>     var result = await ctx.Set<GetServicePlanConfigurationByCurrencyResult>().FromSqlRaw("select 1").ToListAsync();
> ```

I need syncrhonous call (its a long story related to legacy)

**Comment by ErikEJ** [comment]
You can just use ToList() then.

**Comment by Belorus** [comment]
Good, that's clear.

Sorry for silly Q, just to understand it better:

The strategy in EF Core is to have 2 separate implementations (sync + async) and usage of async in sync way in .NET Framework is not supported (due to lack of .ConfigureAwait(false))

**Comment by davidfowl** [comment]
This isn't adding up:
- Is this an ASP.NET application running on .NET Framework?
- Why aren't you using ToList instead of ToListAsync?

**Comment by roji** [comment]
> The strategy in EF Core is to have 2 separate implementations (sync + async) and usage of async in sync way in .NET Framework is not supported (due to lack of .ConfigureAwait(false))

It's worth nothing that this has nothing to do with EF Core. In .NET in general, you should always avoid synchronously blocking on an asynchronous operation (this is known as sync-over-async, [read e.g. this](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html)).

**Comment by raman-kazhadub** [comment]
> > The strategy in EF Core is to have 2 separate implementations (sync + async) and usage of async in sync way in .NET Framework is not supported (due to lack of .ConfigureAwait(false))
> 
> It's worth nothing that this has nothing to do with EF Core. In .NET in general, you should always avoid synchronously blocking on an asynchronous operation (this is known as sync-over-async, [read e.g. this](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html)).
@davidfowl @roji 

one point from this article:
> In your “library” async methods, use ConfigureAwait(false) wherever possible.

ToArrayAsync contains ConfigureAwait(false):
https://github.com/dotnet/efcore/blob/main/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L2244
ToListAsync doesn't:
https://github.com/dotnet/efcore/blob/main/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L2209

Also there is the same problem for ExecuteSqlRawAsync (EF Core 3.1.25) because ConfigureAwait(false) is missing:
https://github.com/dotnet/efcore/blob/v3.1.25/src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs#L656

As I can see, @roji fixed it for EF Core 5,6,7 (https://github.com/dotnet/efcore/pull/21110/commits/645a889460d781d3c2c8d2b299488554eb8f7d4e), but not for 3.1.* (latest available for .NET Framework).

So there are bugs for ToListAsync and ExecuteSqlRawAsync in EF Core 3.1.25 because ConfigureAwait(false) missing.

**Comment by roji** [comment]
@raman-kazhadub EF Core 3.1 didn't have ConfigureAwait(false) internally, by design; EF Core may call into user code (i.e. when materializing entities), and at the time we thought it was important to maintain the user's synchronization context for data binding scenarios (i.e. in case the entity property being set calls into a UI action). We changed this in EF Core 5, but still avoid doing `ConfigureAwait(false)` in the specific code flow of SaveChanges for the above reason.

In other words, the guidance to use `ConfigureAwait(false)` is definitely right, but isn't absolute - there are some cases where exceptions makes sense.

In any case, this again isn't directly related to the guidance against synchronously blocking on asynchronous code, as you're doing above. Even with a library that contains `ConfigureAwait(false)` everywhere, doing this could cause thread pool starvations and is a very bad idea.

**Comment by Belorus** [comment]
Think of that:

* The best codegen by @ErikEJ doesn't support sync mode (and no plans) https://github.com/ErikEJ/EFCorePowerTools/issues/1417
* EF Core 3.1 is the latest version that can be used in .NET Framework
* EF Core 3.1 doesn't support sync via async mode
* EF 5.0 **does** has fix that makes sync via sync possible, altho it is never used in legacy projects

Migration becomes a bit hard in such circumstances. When talking about big projects it always good to be able to move step by step: migrate to EF Core while on Framework, do sync via async, and then slowly refactor code, endpoint by endpoint.

If we should make everything async in 1 commit - that is risky. 
If we should fork amazing codegen by@ErikEJ to add sync mode - that is also not a great way, since he doesn't have such plans (and i fully understand him)

Maybe backporting your commit to 3.1 makes sense?

**Comment by ErikEJ** [comment]
@Belorus you made your case by flattering me. Maybe we could investigate togethet what it would take to add sync support to sproc code generation?

**Comment by Belorus** [comment]
> @Belorus you made your case by flattering me. Maybe we could investigate togethet what it would take to add sync support to sproc code generation?

Haha, to be honest, we would be happy to make a contribution to your library. If you could find time to review the PR and give any recommendations - that would be best help.

Just give us (me and Roman) green light, and will return back with some code

**Comment by roji** [comment]
> EF Core 3.1 doesn't support sync via async mode

I'll repeat - sync-over-async is discouraged *always* - regardless of `ConfigureAwait(false)` - so this isn't an EF Core thing. Even with EF Core 3.1, doing sync-over-async can cause various starvation issues and problems. Do not do this.

> When talking about big projects it always good to be able to move step by step: migrate to EF Core while on Framework, do sync via async, and then slowly refactor code, endpoint by endpoint.

That is a specifically-discouraged practice: you should not gradually convert an application by using async in certain places and then doing sync-over-async over them. See this particular point in [Async all the way](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming?WT.mc_id=DT-MVP-5000058#async-all-the-way).

What you can do, is gradually convert sync to async "vertically", i.e. ensure that complete call stacks are fully async, but leaving others sync (as long as they don't interact). That allows a gradual transition.

> Maybe backporting your commit to 3.1 makes sense?

Adding `ConfigureAwait(false)` everywhere is a big change that could have various consequences; it isn't safe to backport such a thing to a 3.1 patch version.

**Comment by ErikEJ** [comment]
@Belorus See https://github.com/ErikEJ/EFCorePowerTools/issues/1417 for further discussion

**Comment by lkomanetz** [comment]
@roji @davidfowl 
Is it possible to have `ToListAsync` and `ToArrayAsync` return the task that can be configured by the caller for their scenario?  I believe that's how it worked in EF6 where there was no asynchronous work happening within the method body.  I can understand that because it is streaming the contents asynchronously in order to create the `List<T>` you need to asynchronously wait.  I'm just curious if there is a way to write it so the `Task<List<T>>` is provided and it can be configured or if the async stream provides more benefit than providing that task.

Maybe an overload like this?
```cs
public static async Task<List<TSource>> ToListAsync<TSource>(
    this IQueryable<TSource> source,
    bool continueOnCapturedContext = true,
    CancellationToken cancellationToken = default)
{
    var list = new List<TSource>();
    await foreach (var element in source.AsAsyncEnumerable().WithCancellation(cancellationToken).ConfigureAwait(continueOnCapturedContext))
    {
        list.Add(element);
    }

   return list;
}
```

**Comment by roji** [comment]
@lkomanetz you're asking why ToListAsync doesn't internally do ConfigureAwait(false), right? If so, that's a bug - it should always do that (i.e. we shouldn't expose a new overload which allows deciding whether it should). I've just opened #32449 to track it.
